src/Services/Virtualization/PortalAuthenticationModule.cs

```diff
@@ -20,158 +21,80 @@
         }
 
-        public void Init(HttpApplication context)
+        public void Init(HttpApplication application)
         {
             FormsAuthentication.Initialize();
-            context.AuthenticateRequest += new EventHandler(OnAuthenticateRequest);
-            context.EndRequest += new EventHandler(OnEndRequest); // Forms
-            context.AuthorizeRequest += new EventHandler(OnAuthorizeRequest);
+            application.AuthenticateRequest += OnAuthenticateRequest;
+            application.EndRequest += OnEndRequest; // Forms
         }
 
-        private void OnAuthorizeRequest(object sender, EventArgs e)
+        private const string AccessSignatureName = "as";
+        private const string RefreshSignatureName = "rs";
+        private const string AccessHeaderName = "X-Access-Data";
+        private const string RefreshHeaderName = "X-Refresh-Data";
+        private const string AuthenticationTypeHeaderName = "X-Authentication-Type";
+        private const string TokenLoginPath = "/sn-token/login";
+        private const string TokenRefreshPath = "/sn-token/refresh";
+        private static class HttpResponseStatusCode
         {
-            OnAuthorizeRequest((sender as HttpApplication)?.Context);
+            public static int Unauthorized = 401;
+            public static int Ok = 200;
         }
-        internal void OnAuthorizeRequest(HttpContext context)
-        {
-            PortalContext currentPortalContext = PortalContext.Current;
-            if (currentPortalContext == null)
-                return;
-            var currentUser = context?.User.Identity as User;
-
-            // deny access for visitors in case of webdav or office protocol requests, if they have no See access to the content
-            if (currentUser != null && currentUser.Id == Identifiers.VisitorUserId && (currentPortalContext.IsOfficeProtocolRequest || currentPortalContext.IsWebdavRequest))
-            {
-                if (!currentPortalContext.IsRequestedResourceExistInRepository ||
-                    currentPortalContext.ContextNodeHead == null ||
-                    !SecurityHandler.HasPermission(currentPortalContext.ContextNodeHead, PermissionType.See))
-                {
-                    AuthenticationHelper.ForceBasicAuthentication(HttpContext.Current);
-                }
-            }
+        private static ISecurityKey _securityKey;
+        private static object _keyLock = new object();
 
-            if (context == null)
-                return;
-
-            if (currentPortalContext.IsRequestedResourceExistInRepository)
+        private ISecurityKey SecurityKey
+        {
+            get
             {
-                var authMode = currentPortalContext.AuthenticationMode;
-                if (string.IsNullOrEmpty(authMode))
-                    authMode = WebApplication.DefaultAuthenticationMode;
-
-                bool appPerm;
-                if (authMode == "Forms")
+                lock (_keyLock)

```

Comment:
This means unnecessary locks every time. The outer 'if' is missing from the if-lock-if pattern.